### PR TITLE
New pixel_type constants and some affected base functions

### PR DESCRIPTION
**Pull request by pinterf**

### CODE REVIEW

**Comment by pinterf** [comment]
Work in progress, I have problems implementing the 4th plane for YUVA and RGBA

**Comment by chikuzen** [comment]
Is the 4th plane really necessary?
I think alpha should be processed on packed format (or independent gray frame like vapoursynth).

**Comment by pinterf** [comment]
Ok, 4th plane handling is OK now.
But it is clear that none of the internal filters will work on YUVA and planar RGBA (BGRA), simply too much work. Making two clips, one w/o alpha info, and one Alpha clip, is a viable solution.
Source filters could create such frames and one could split them with a minimal number of supporting scipt functions
e.g. 
clip=whateversource(...,"GBRAP10") #or "YUVAP10" #4 planes
clip_3planes_no_alpha=StripAlpha(clip) # we get 3-plane planar (RGBP or YUV)
clip_greyscale_alpha = AlphaToY(clip)
and when exporting
clip = MergeWithAlpha(clip_3planes_no_alpha,clip_greyscale_alpha)

Opinions?

**Comment by pinterf** [comment]
Updated pixel_type description in avisynth.h
Constants for planar video format with alpha channel and code for frame creation is done.
VideoInfo and script functions: Is420, Is422, Is444, IsYUVA, IsRGBA, IsPlanarRGB (also for the c interface)
The existing IsYUV and IsRGBxx functions are returning true only for the old YUV and packed RGB formats.

File: avs_core/include/avisynth_c.h [hunk 1]
```diff
@@ -259,26 +288,26 @@ typedef struct AVS_VideoInfo {
...
 
-AVSC_INLINE int avs_is_rgb24(const AVS_VideoInfo * p) 
-        { return (p->pixel_type&AVS_CS_BGR24)==AVS_CS_BGR24; } // Clear out additional properties
+AVSC_INLINE int avs_is_rgb24(const AVS_VideoInfo * p)
+        { return ((p->pixel_type&AVS_CS_BGR24)==AVS_CS_BGR24) && ((p->pixel_type & AVS_CS_SAMPLE_BITS_MASK) == AVS_CS_SAMPLE_BITS_8); }
 
-AVSC_INLINE int avs_is_rgb32(const AVS_VideoInfo * p) 
-        { return (p->pixel_type & AVS_CS_BGR32) == AVS_CS_BGR32 ; }
+AVSC_INLINE int avs_is_rgb32(const AVS_VideoInfo * p)
+       { return ((p->pixel_type&AVS_CS_BGR24)==AVS_CS_BGR32) && ((p->pixel_type & AVS_CS_SAMPLE_BITS_MASK) == AVS_CS_SAMPLE_BITS_8); }
```
Comment (qyot27) [inline avs_core/include/avisynth_c.h]: Typo?

Comment (pinterf) [inline avs_core/include/avisynth_c.h]: Yes, thank you

**Comment by pinterf** [comment]
YUVA and Planar RGBA test script. (PlaneToY and BlankClip demo)
Left clips should be equal (rgb color = FF1493). Right clips are alphas, top=dark($60) bottom:light($A0)

rgbap=BlankClip(width=320, height=200, color    =$60FF1493, pixel_type="RGBAP")
yuva =BlankClip(width=320, height=200, color_yuv=$80FF1493, pixel_type="YUVA444")
argb =BlankClip(width=320, height=200, color    =$A0FF1493, pixel_type="RGB32")

r=rgbap.PlaneToY("R")
g=rgbap.PlaneToY("G")
b=rgbap.PlaneToY("B")
a=rgbap.PlaneToY("A")
argb_from_rgbpa=MergeARGB(a,r,g,b)

ry=yuva.PlaneToY("Y") #or ry=yuva.ConvertToY()
gy=yuva.PlaneToY("U")
by=yuva.PlaneToY("V")
ay=yuva.PlaneToY("A")
argb_from_yuva=MergeARGB(ay,ry,gy,by)

StackHorizontal(\
Stackvertical(argb_from_rgbpa.SubTitle("source=rgbpa"),\
              argb_from_yuva.SubTitle("source=yuva"),\
              argb.SubTitle("source=rgb32")),\
StackVertical(ConvertToRGB32(a).SubTitle("Alpha from rgbpa"),\
              ConvertToRGB32(ay).SubTitle("Alpha from yuva"),\
              argb.ShowAlpha().SubTitle("Alpha from rgb32"))\
)

**Comment by pinterf** [comment]
Test scripts for 16 bit MergeRGB/ARGB

rgbap=BlankClip(width=320, height=200, color    =$60FF1493, pixel_type="RGBAP")
rgbap16=BlankClip(width=320, height=200, color  =$60FF1493, pixel_type="RGBAP16")
argb =BlankClip(width=320, height=200, color    =$A0FF1493, pixel_type="RGB32")
rgb =BlankClip(width=320, height=200, color    =$A0FF1493, pixel_type="RGB24")

r16=rgbap16.PlaneToY("R")
g16=rgbap16.PlaneToY("G")
b16=rgbap16.PlaneToY("B")
a16=rgbap16.PlaneToY("A")
argb_from_rgbap16=MergeARGB(a16,r16,g16,b16).ConvertTo8bit()
rgb_from_rgbap16=MergeRGB(r16,g16,b16).ConvertTo8bit()
rgb24_from_rgbap16=MergeRGB(r16,g16,b16,pixel_type="RGB48").ConvertTo8bit()

r=rgbap.PlaneToY("R")
g=rgbap.PlaneToY("G")
b=rgbap.PlaneToY("B")
a=rgbap.PlaneToY("A")
argb_from_rgbap=MergeARGB(a,r,g,b)
rgb_from_rgbap=MergeRGB(r,g,b)
rgb24_from_rgbap=MergeRGB(r,g,b,pixel_type="RGB24")

/* # Test A: MergeARGB
Stackvertical(argb_from_rgbap.SubTitle("source=rgbpa"),\
              argb_from_rgbap16.SubTitle("source=rgbap16"),\
              argb.SubTitle("source=rgb32"))
*/
# Test B: MergeRGB no pixel_type-> defaults to RGB32 or RGB64
Stackvertical(rgb_from_rgbap.SubTitle("source=rgbpa"),\
              rgb_from_rgbap16.SubTitle("source=rgbap16"),\
              argb.SubTitle("source=rgb24"))

/* # Test C: MergeRGB pixel_type param RGB24 or RGB48
Stackvertical(rgb24_from_rgbap.SubTitle("source=rgbpa"),\
              rgb24_from_rgbap16.SubTitle("source=rgbap16"),\
              rgb.SubTitle("source=rgb24"))
*/

**Comment by pinterf** [comment]
Test cases for BlackClip RGB48 and RGB32

a8=BlankClip(width=320, height=200, color =$A0FF1493, pixel_type="RGB24")
a16=BlankClip(width=320, height=200, color=$A0FF1493, pixel_type="RGB48").ConvertTo8bit()
StackVertical(a8,a16)

a8=BlankClip(width=320, height=200, color =$A0FF1493, pixel_type="RGB32")
a16=BlankClip(width=320, height=200, color=$A0FF1493, pixel_type="RGB64").ConvertTo8bit()
StackVertical(a8,a16)

**Comment by pinterf** [comment]
ConvertTo16Bit() have a new bitdepth parameter, defaulting to 16. 
e.g. ConvertTo16Bit(bitdepth=10) will make a YUV444P10 from a YV24 clip.
Preliminary non-working "int modifyrange" parameter e.g. if real 10 bit range is instead of 16 bit.
ConvertTo16Bit(bitdepth=10, modifyrange=true) 

test script
rgbap16=BlankClip(width=320, height=200, color    =$60FF1493, pixel_type="RGBAP").ConvertTo16bit(bitdepth=10).Info().ConvertTo16bit(bitdepth=16)
rgbap=BlankClip(width=320, height=200, color  =$60FF1493, pixel_type="RGBAP16").Info().ConvertTo8Bit()
argb =BlankClip(width=320, height=200, color    =$A0FF1493, pixel_type="RGB32").ConvertTo16Bit().Info().ConvertTo8bit()
rgb =BlankClip(width=320, height=200, color    =$A0FF1493, pixel_type="RGB24").ConvertTo16Bit().Info().ConvertTo8bit()


r16=rgbap16.PlaneToY("R")
g16=rgbap16.PlaneToY("G")
b16=rgbap16.PlaneToY("B")
a16=rgbap16.PlaneToY("A")
argb_from_rgbap16=MergeARGB(a16,r16,g16,b16).ConvertTo8bit()
rgb_from_rgbap16=MergeRGB(r16,g16,b16).ConvertTo8bit()
rgb24_from_rgbap16=MergeRGB(r16,g16,b16,pixel_type="RGB48").ConvertTo8bit()

r=rgbap.PlaneToY("R")
g=rgbap.PlaneToY("G")
b=rgbap.PlaneToY("B")
a=rgbap.PlaneToY("A")
argb_from_rgbap=MergeARGB(a,r,g,b)
rgb_from_rgbap=MergeRGB(r,g,b)
rgb24_from_rgbap=MergeRGB(r,g,b,pixel_type="RGB24")

 # Test A: MergeARGB
/*
Stackvertical(argb_from_rgbap.SubTitle("source=rgbpa"),\
              argb_from_rgbap16.SubTitle("source=rgbap16"),\
              argb.SubTitle("source=rgb32"))
*/
# Test B: MergeRGB no pixel_type-> defaults to RGB32 or RGB64
Stackvertical(rgb_from_rgbap.SubTitle("source=rgbpa"),\
              rgb_from_rgbap16.SubTitle("source=rgbap16"),\
              argb.SubTitle("source=rgb24"))

/* # Test C: MergeRGB pixel_type param RGB24 or RGB48
Stackvertical(rgb24_from_rgbap.SubTitle("source=rgbpa"),\
              rgb24_from_rgbap16.SubTitle("source=rgbap16"),\
              rgb.SubTitle("source=rgb24"))
*/

**Comment by qyot27** [comment]
Do the changes to avs_is_rgb24/32 and avs_is_yuy2 (or using the *_GENERIC_* constants in the definitions for the older 2.6 pixfmts) in the interface header(s) constitute a breaking API change?  If they do, the AVISYNTH_INTERFACE_VERSION definitely needs to be bumped and/or this may be a good time to introduce something like an AVSPLUS_INTERFACE_VERSION so a receiving application can check specifically for AviSynth+ vs. AviSynth 2.6.

**Comment by pinterf** [comment]
I don't think so, only enums were added for better readibility. But the usage of new VideoInfo functions IsPlanarRGB, etc... will of course fail on classic avs.

**Comment by chikuzen** [comment]
IIRC, Stack16 formats are always 16bit.
ConvertFromStacked requires (16 -bits) shift to right, doesn't. it?

**Comment by qyot27** [comment]
>IIRC, Stack16 formats are always 16bit.
ConvertFromStacked requires (16 -bits) shift to right, doesn't. it?

f3kdb can output 10, 12, and 14 in Stacked format too.  Dunno if Dither can.


Question: when using avs_is_444 (or the other two) for detecting either plain or alpha-enabled formats, and then using that information to allocate enough room for them, it is enough to do `avs_bits_per_component(avs->vi) * 4`, correct?

It works fine for either 3 or 4 planes, obviously, but will having it allot room for 4 planes create inefficient overhead for only 3?  Would it be worth it to have a function like `avs_has_alpha_channel` that can report back true/false about whether PLANAR_A is present in the format, and then use that to only allocate 4 if the format has alpha?

See [here](https://github.com/qyot27/FFmpeg/blob/avs_16bit/libavformat/avisynth.c#L658) for what I'm referring to.

File: plugins/ConvertStacked/ConvertStacked.cpp [hunk 2]
```diff
@@ -134,13 +134,41 @@ class ConvertFromStacked : public GenericVideoFilter
...
+            vi.pixel_type = VideoInfo::CS_Y14;
+        else if (bits == 16 && vi.IsYV12())
+            vi.pixel_type = VideoInfo::CS_YUV420P16;
+        else if (bits == 16 && vi.IsYV16())
+            vi.pixel_type = VideoInfo::CS_YUV422P16;
+        else if (bits == 16 && vi.IsYV24())
+            vi.pixel_type = VideoInfo::CS_YUV444P16;
+        else if (bits == 16 && vi.IsY8())
+            vi.pixel_type = VideoInfo::CS_Y16;
+        else env->ThrowError("ConvertStackedToNative: Input stacked clip must be YV12, YV16, YV24 or Y8");
```
Comment (qyot27) [inline plugins/ConvertStacked/ConvertStacked.cpp]: Whoops, looks like an earlier version slipped through here when I updated the patch.  That needs to be changed to ConvertFromStacked as well.

Comment (pinterf) [inline plugins/ConvertStacked/ConvertStacked.cpp]: 1.) You have an is_yuva() function. If true, the number of planes is 4.
Or call vi.ComponentCount. It will return 3 for Planar RGB and YUV, and 4 for Planar RGBA or YUVA.
2.) I can see a serious confusion on bits_per_component usage.
When you ask for 8 bit types it returns 8. For 16 bit types it returns 10-12-14-16.
In the linked ffmpeg source bits_per_component is used for yuv420 types. For 8 bit the function returns bits=8. For YV12 the result is 3x8=24. For a YV16 or YV24 it will also return 3*8=24

The "harcoded" avs 2.6 branch calls the classic "bits_per_pixel" function which is a special function that returns the _effective_ _average_ bit accupation for whole pixel. Let's see a YV12 (y-u-v) pixel. Classic bitsPerPixel returns 12 (reason: for 4 pixels = 4 bytes luma + 2 byte chroma = 6 byte / 4 pixel -> 1.5 byte / 1 pixel -> 12 bits for a pixel.)

You have to decide what that bits really means here. I can see that a stream-like packet size is calculated from "bits".
I recommend that the Classic avs bitsperPixel calculation should be used then additionally be scaled with *bits_per_component/8.

That also implies that a 10 bit avisynth format contains real 10 bit data and not scaled to 16 (?).

3.) Then later comes when rgb is checked. Planar RGB is also top bottom?

That's all for now, I will possibly off for a week from now (holiday).

**Comment by chikuzen** [comment]
>f3kdb can output 10, 12, and 14 in Stacked format too.

L-SMASH-Works and VapourSource, too.
But those are able to output frames as interleaved(double width) formats, too.
Why output as stacked from f3kdb?
Double width output is faster than stacked, and ConvertFromDoubleWidth() is zero cost.

**Comment by qyot27** [comment]
>Why output as stacked from f3kdb?

I really don't know.  I never use stacked format myself, it was merely the first one that could be bitshifted back together inside AviSynth+ and output as real 16-bit.  It was a really short window (until ConvertFromDoubleWidth was added) where I cared about it at all.

File: plugins/ConvertStacked/ConvertStacked.cpp (hunk 2 above)
Comment (qyot27) [inline plugins/ConvertStacked/ConvertStacked.cpp]: On 1.), I don't know why I didn't think of using avs_num_components to determine whether 3 or 4 planes were needed.

Point 2.) requires some background. The bit allocations you see under the 2.6 part are wonky as it is.  The real story is that the AviSynth demuxer in libavformat was initially rewritten by (one of?) AvxSynth's developers, and since AvxSynth was essentially just AviSynth 2.5 with some of 2.6's API backported to it, the plain avs_bits_per_pixel use meant that it worked for YV12 because it also worked for AviSynth 2.5, and 2.6 hadn't changed anything in that regard.  The rest of the formats being parsed were all packed (YUY2, RGB24, RGB32), so bits_per_pixel returned the correct values.  When I was helping bug test the rewritten demuxer to get it committed, I discovered that you could get 2.6's newer planar formats (YV24, YV16, YV411, and Y8) to work if the actual bpp value was explicitly defined, because avs_bits_per_pixel didn't work with them.

And these explicitly-defined values worked for us, too, even when the high bit depths were added (earlier versions of the FFmpeg patch could open AVS_CS_YUV444P16 and recognize it as yuv444p16 without messing with the bit allocations at all).  It might still work against the current upstream MT, but not against this pull request.  Whatever change occurred to vi.BitsPerPixel in AviSynth+ (or this pull request specifically), YV12 stopped working with avs_bits_per_pixel, causing FFmpeg to crash, but with it explicitly defined to a bit value like the rest of the planar formats, it started working again.  Whatever the change was, it also made the high bit depths stop working with avs_bits_per_pixel.  As above, though, the packed ones were unaffected.  RGB24/32, RGB48/64, and YUY2 were detected correctly.

Basically I just said "screw it" and used avs_bits_per_component since it's dumb math and libavformat doesn't complain about packets being too big for a sample.  I'm not opposed to making it fit the pix_fmts more exactly, but since FFmpeg still supports 2.6, it has to be shielded from the avsplus-exclusive functions, so it'll still be branched like it currently is.  It's probable that `avs_bits_per_pixel * avs_bits_per_component / 8` would have been suggested if I'd submitted the current form of the patch to their mailing list anyway, so it doesn't hurt anything to correct it now.

With 3.), I just hadn't gotten around to adjusting that.  Packed RGB has to be flipped, and it simply doesn't check for planar RGB yet.  It does need flipping too.

Comment (qyot27) [inline plugins/ConvertStacked/ConvertStacked.cpp]: After testing, calling avs_bits_per_pixel on any planar format results in a crash, so the proposed `avs_bits_per_pixel * avs_bits_per_component / 8` doesn't work.  However, I was able to cut down the current avsplus-specific method to a much simpler form, with YV411 being the only oddball.

Comment (pinterf) [inline plugins/ConvertStacked/ConvertStacked.cpp]: BitsPerpixel was not intended to change. That's why BitsPerComponent was introduced.
Nevertheless BitsPerPixel now works for YUVA as for YUV. See  in Pull Request.
It should return 12 for YV12, 16 for YV16, etc... Same as in classic avisynth. If it returns these values, the crash reason should be searched at somewhere else.

Comment (qyot27) [inline plugins/ConvertStacked/ConvertStacked.cpp]: The results of what avs_bits_per_pixel reports:
>RGB24: 24
RGB32: 32
RGB48: 48
RGB64: 64
YUY2: 16
YV24: 03
YV12: 01
YV16: 02
YV411: 01
Y8: 08
Y16: 16
YUV444P16: 06

Compare this to AviSynth 2.6 (I took the hardcoded values away and used only avs_bits_per_pixel, as I was starting to suspect it wasn't really necessary either):
>YV24: 24
YV12: 12
YV16: 16
YV411: 12

So yeah, BitsPerPixel isn't reporting the right bit values for planar formats (it looks like it's reporting *byte* values instead).  Y8 and Y16 would be fine, and so would all the packed formats, but the rest aren't correct, and trying to undo that in libavformat is both A) not the right way of fixing it and B) would require if/else-ing it so that the planar and packed formats don't screw each other up.
